### PR TITLE
Fix Boruvka dendrogram generation

### PR DESCRIPTION
**Pull request by aprokop**
Fix cases where a dendrogram could become non-binary due to same weight edges

### CODE REVIEW

File: test/tstDendrogram.cpp [hunk 1]
```diff
@@ -188,4 +188,53 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(dendrogram_boruvka, DeviceType,
...
+  using Point = ArborX::ExperimentalHyperGeometry::Point<2, float>;
+
+  ExecutionSpace space;
+
+  int const N = 101;
+  int const n = N * N;
+  std::vector<Point> points_v(N * N);
+  for (int i = 0; i < N; ++i)
+    for (int j = 0; j < N; ++j)
+      points_v[(2 * i % N) * N + (2 * j % N)] = Point{(float)i, (float)j};
```
Comment (masterleinad) [inline test/tstDendrogram.cpp:207]: Doesn't that mean you don't initialize odd values here and overwrite even values?
Can you please add a comment explaining why/how you do the initialization here?

File: src/details/ArborX_MinimumSpanningTree.hpp [hunk 2]
```diff
@@ -620,6 +621,54 @@ void computeParents(ExecutionSpace const &space, Edges const &edges,
...
+            ++k;
+            if (edges(permute(k)) < edges(permute(m)))
+              m = k;
+          }
+
+          // Place the smallest edge at the beginning of the chain
+          if (m != i)
+            KokkosExt::swap(permute(i), permute(m));
+        }
+      });
```
Comment (masterleinad) [inline src/details/ArborX_MinimumSpanningTree.hpp:670]: How expensive is this? Is the runtime negligible to the other kernels we are running?

Comment (aprokop) [reply src/details/ArborX_MinimumSpanningTree.hpp:670]: It is negligible in practice.

File: test/tstDendrogram.cpp [hunk 3]
```diff
@@ -188,4 +188,56 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(dendrogram_boruvka, DeviceType,
...
+  // All points (except border ones) will have the same core distance.
+  int const N = 4;
+  int const n = N * N;
+  std::vector<Point> points_v(N * N);
+  for (int j = 0; j < N; ++j)
+    for (int i = 0; i < N; ++i)
+      points_v[j * N + i] = Point{(float)i, (float)j};
+  auto points = ArborXTest::toView<DeviceType>(points_v, "Testing::points");
+
+  // minpts = 5 is the first value that the test fails with N = 4
```
Comment (masterleinad) [inline test/tstDendrogram.cpp:212]:
```suggestion
  // minpts = 5 is the first value that the test failed with N = 4
```

File: src/details/ArborX_MinimumSpanningTree.hpp [hunk 4]
```diff
@@ -612,13 +613,62 @@ void computeParents(ExecutionSpace const &space, Edges const &edges,
           key = INT_MAX;
 
         // Comparison of weights as ints is the same as their comparison as
-        // floats as long as they are positive and are not NaNs or inf
+        // floats as long as they are positive and are not NaNs or inf.
         static_assert(sizeof(int) == sizeof(float));
-        keys(e) = (key << shift) + KokkosExt::bit_cast<int>(edge.weight);
+        keys(e) = (key << shift) +
+                  (edge.weight ? KokkosExt::bit_cast<int>(edge.weight) : 0);
```
Comment (dalg24) [inline src/details/ArborX_MinimumSpanningTree.hpp:619]: Wait what?
isn't `bit_cast<int>(0.f)` equal to `0`?

Comment (dalg24) [reply src/details/ArborX_MinimumSpanningTree.hpp:619]: https://godbolt.org/z/93ssn1Tf3

Comment (aprokop) [reply src/details/ArborX_MinimumSpanningTree.hpp:619]: I thought maybe to be a bit more careful, as the bit cast of a negative 0 is not 0. But that shouldn't happen in practice, unless someone does some weird distance calculations. 

Would you prefer to change it back?

File: src/details/ArborX_MinimumSpanningTree.hpp [hunk 5]
```diff
@@ -612,13 +613,62 @@ void computeParents(ExecutionSpace const &space, Edges const &edges,
...
+          // Find m, the index of the smallest edge with the same weight in
+          // this chain
+          int k = i;
+          int m = i;
+          while (k < (int)num_edges - 1 && keys(k + 1) == key)
+          {
+            ++k;
+            if (edges(permute(k)) < edges(permute(m)))
+              m = k;
+          }
```
[the remark this answers is not in the document]
Comment (aprokop) [reply src/details/ArborX_MinimumSpanningTree.hpp]: You are right, with a fix. Changed.

File: src/details/ArborX_MinimumSpanningTree.hpp [hunk 6]
```diff
@@ -612,13 +615,61 @@ void computeParents(ExecutionSpace const &space, Edges const &edges,
           key = INT_MAX;
 
         // Comparison of weights as ints is the same as their comparison as
-        // floats as long as they are positive and are not NaNs or inf
+        // floats as long as they are positive and are not NaNs or inf. We use
+        // signbit instead of >= 0 just as an extra precaution against negative
+        // floating zeros.
         static_assert(sizeof(int) == sizeof(float));
+        assert(Kokkos::isfinite(edge.weight) &&
```
Comment (masterleinad) [inline src/details/ArborX_MinimumSpanningTree.hpp:622]:
```suggestion
        KOKKOS_ASSERT(Kokkos::isfinite(edge.weight) &&
```

File: src/details/ArborX_MinimumSpanningTree.hpp [hunk 7]
```diff
@@ -612,13 +615,61 @@ void computeParents(ExecutionSpace const &space, Edges const &edges,
...
+        auto key = keys(i);
+
+        if (i == 0 || ((keys(i - 1) >> shift) != (key >> shift)))
+        {
+          // i is at the start of a chain
+
+          // Find the index of the smallest edge with the same weight in
+          // this chain
+          int m = i;
+          for (int k = i + 1; k < (int)num_edges && keys(k) == key; ++k)
```
Comment (masterleinad) [inline src/details/ArborX_MinimumSpanningTree.hpp:663]: You could also just define `num_edges` as an `int`.

**Review by masterleinad — APPROVED** [review]

File: src/details/ArborX_MinimumSpanningTree.hpp [hunk 8]
```diff
@@ -24,6 +25,14 @@
 #include <ArborX_HyperBox.hpp>
 #include <ArborX_LinearBVH.hpp>
 
+#if KOKKOS_VERSION >= 40200
+#include <Kokkos_Assert.hpp> // KOKKOS_ASSERT
+#else
+#include <impl/Kokkos_Error.hpp> // KOKKOS_ASSERT
+#endif
+
+#include <Kokkos_MathematicalFunctions.hpp> // isfinite, signbit
```
Comment (dalg24) [inline src/details/ArborX_MinimumSpanningTree.hpp:34]: There is no point, we include `<Kokkos_Core.hpp>` below.

File: test/tstDendrogram.cpp [hunk 9]
```diff
@@ -188,4 +188,56 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(dendrogram_boruvka, DeviceType,
...
+      Kokkos::view_alloc(space, "Testing::count"), 2 * n - 1);
+
+  Kokkos::parallel_for(
+      "Testing::count_children",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, 2 * n - 1),
+      KOKKOS_LAMBDA(int i) {
+        Kokkos::atomic_inc(&counts(mst.dendrogram_parents(i)));
+      });
+
+  int wrong_counts = 0;
```
Comment (dalg24) [inline test/tstDendrogram.cpp:229]:
```suggestion
  int wrong_counts;
```

File: src/details/ArborX_MinimumSpanningTree.hpp (hunk 4 above)
Comment (dalg24) [reply src/details/ArborX_MinimumSpanningTree.hpp]: I don't fully get it.  Let's talk.

**Pull request merged**